src/proto/h2/client.rs

```diff
@@ -280,25 +281,20 @@ where
     fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
         let mut this = self.project();
 
-        if let Some(conn) = this.conn {
-            conn.poll_unpin(cx).map(|_| ())
-        } else {
-            match ready!(this.select.poll_unpin(cx)) {
-                Either::Left((_, _)) => {
-                    // ok or err, the `conn` has finished
-                    return Poll::Ready(());
-                }
-                Either::Right((_, b)) => {
-                    // mpsc has been dropped, hopefully polling
-                    // the connection some more should start shutdown
-                    // and then close
-                    trace!("send_request dropped, starting conn shutdown");
-                    drop(this.cancel_tx.take().expect("Future polled twice"));
-                    this.conn = &mut Some(b);
-                    return Poll::Pending;
-                }
-            }
-        }
+        if let Poll::Ready(_) = this.conn.poll_unpin(cx) {
```

Comment:
As far as I'm aware, fusing a future doesn't allow it to be polled multiple times. You have to explicitly check whether the future is already terminated using the [`is_terminated()`](https://docs.rs/futures/latest/futures/future/trait.FusedFuture.html#tymethod.is_terminated) method.

Reply:
Thanks for the feedback. My understanding on this is different, but maybe I read it wrong.

The outer future here (and below too) is a `Fuse` (see https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.fuse) which internally already takes care of shunting subsequent polls to `Pending`:
https://github.com/rust-lang/futures-rs/blob/89c06f5e89b8499102a198736ccc859949e049a3/futures-util/src/future/future/fuse.rs#L82-L90
Did I misunderstand something here?

Or maybe are you suggesting to avoid introducing two `Fuse` struct wrappers here and directly try to implement the `FusedFuture` trait on the two existing underlying futures (and then check `is_terminated()`)?

Reply:
> Thanks for the feedback. My understanding on this is different, but maybe I read it wrong.
The outer future here (and below too) is a Fuse (see https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.fuse) which internally already takes care of shunting subsequent polls to Pending:
https://github.com/rust-lang/futures-rs/blob/89c06f5e89b8499102a198736ccc859949e049a3/futures-util/src/future/future/fuse.rs#L82-L90
Did I misunderstand something here?
Or maybe are you suggesting to avoid introducing two Fuse struct wrappers here and directly try to implement the FusedFuture trait on the two existing underlying futures (and then check is_terminated())?


I agree. As far as i understand the `Future` implementation of `Fuse`, it is already checked in the `poll` method.

Reply:
I pushed a commit on top which gets rid of additional `Fuse` structs and directly uses the `FusedFuture` trait (also found out that `StreamFuture` already implements that).

The two versions should be functionally equivalent, so I'm happy to keep whatever form is preferred by the maintainers.